File: sources/scala/tools/scalac/transformer/matching/RightTracerInScala.scala
/**
 *  $Id$
 */

import scalac._;
import scalac.ast._;
import scalac.symtab._;
import Tree._;

import java.util._ ;
import Scope.SymbolIterator;

import scalac.util.Name ;
import scalac.util.Names ;

import scala.tools.util.Position;

package scala.tools.scalac.transformer.matching {

  /** translate right tracer to code
     * @param dfa determinized left tracer
     * @param left nondeterm. left tracer (only needed for variables!)
     * @param cf   ...
     * @param pat  ?
     * @param elementType ...
     */
class RightTracerInScala(dfa: DetWordAutom, seqVars: Set, owner: Symbol, cf: CodeFactory, pat:Tree, elementType: Type)
extends TracerInScala( dfa, elementType, owner, cf )  {

  final def defs = cf.defs;

  val allVars: Set = collectVars( pat );

  var varsToExport: Set = new HashSet(); // @todo HANDLE seqVars THESE GLOBALLY INSTEAD OF LOCALLY

  varsToExport.addAll( allVars );
  varsToExport.removeAll( seqVars );

  var targetSym:Symbol = _;

  var helpMap  = new HashMap();
  var helpMap2 = new HashMap();
  var helpVarDefs = new TreeList();

  val it = seqVars.iterator();
  while(it.hasNext()) {
    makeHelpVar( it.next().asInstanceOf[Symbol] );
  }

  val jt = allVars.iterator();
  while(jt.hasNext()) {
    val varSym = jt.next().asInstanceOf[Symbol];
    if(( varSym.name.toString().indexOf("$") == -1 )
       && ( !seqVars.contains( varSym ))) {
         makeHelpVar( varSym, true );
       }
  }

  //System.out.println("allVars: "+allVars);
  //System.out.println("seqVars: "+seqVars);
  //System.out.println("helpVarDefs now: "+helpVarDefs);

  initializeSyms();

  def makeHelpVar(realVar: Symbol): Unit = {
    makeHelpVar( realVar, false );
  }

  /** makes a helpvar and puts mapping into helpMap, ValDef into helpVarDefs
   */

  def makeHelpVar(realVar: Symbol, keepType: Boolean): Unit = {
    val helpVar = owner.newVariable( cf.pos,
                                    0,
                                    cf.fresh.newName( realVar.name
                                                     .toString()+"RTIS" ));
    var rhs: Tree = _;

    //System.out.println("RTiS making helpvar : "+realVar+" -> "+helpVar);

    if( keepType ) {
      helpVar.setType( realVar.getType() );
      rhs = gen.mkDefaultValue( cf.pos, realVar.getType() );
    } else {
      helpVar.setType( defs.LIST_TYPE(elementType) );
      rhs = gen.mkNil( cf.pos );
    }

    helpMap.put( realVar, helpVar );
    helpVar.flags = helpVar.flags | Modifiers.MUTABLE;
    val varDef = gen.ValDef( helpVar, rhs );
    //((ValDef) varDef).kind = Kinds.VAR;
    helpVarDefs.append( varDef );

  }

  def prependToHelpVar(realVar: Symbol, elem:Tree): Tree = {
    val hv = refHelpVar( realVar );
    gen.Assign( hv, gen.mkNewCons( cf.pos, elementType, elem, hv ));
    /*
     return cf.Block(pos,
     new Tree [] {
          cf.debugPrintRuntime( "ASSIGN" ),
          gen.Assign( hv, cf.newSeqCons( elem, hv ))
          }, defs.UNIT_TYPE());
        */
  }

  protected def initializeSyms(): Unit = {

    this.funSym = owner.newLabel( pos,
                                 cf.fresh.newName( "right" ));

    this.iterSym = owner.newVariable( pos,
                                     Modifiers.MUTABLE,
                                     cf.fresh.newName("iter"))
    .setType( cf.SeqTraceType( elementType ));

    this.stateSym = owner.newVariable ( pos,
                                       Modifiers.MUTABLE,
                                       cf.fresh.newName("q"))
    .setType( defs.int_TYPE() ) ;

    this.curSym = owner.newVariable( pos,
                                    Modifiers.MUTABLE,
                                    cf.fresh.newName("cur"))
    .setType( elementType ) ;

    this.targetSym = owner.newVariable( pos,
                                       Modifiers.MUTABLE,
                                       cf.fresh.newName("p"))
    .setType( defs.int_TYPE() ) ;

    funSym.setType( new Type.MethodType( Predef.Array[Symbol] (  // dummy symbol MethodType
      funSym.newVParam( pos, 0, cf.fresh.newName("iter"), cf.SeqTraceType( elementType )),
      funSym.newVParam( pos, 0, cf.fresh.newName( "q" ), defs.int_TYPE()) ),
                                        defs.void_TYPE() )); // result

  }

  // load current elem and trace
  override def loadCurrentElem(body: Tree): Tree = {
    gen.If( cf.isEmpty( _iter() ),
           run_finished( 0 ),            // we are done
           gen.mkBlock( Predef.Array[Tree] (
             gen.ValDef( this.targetSym,
                        cf.SeqTrace_headState( gen.Ident( pos, iterSym))),
             gen.ValDef( this.curSym,
                        cf.SeqTrace_headElem( gen.Ident( pos, iterSym )))),
                       body )
         );
  }

  /** see code_state0_NEW
   */
  def code_state0(elseBody: Tree) = { // careful, map Int to Int

    gen.If( cf.Equals( _state(), gen.mkIntLit( cf.pos, 0 )),
           code_state0_NEW(),
           elseBody );

  }

    /** this one is special, we check the first element of the trace
     *  and choose the next state depending only on the state part
     */
  def code_state0_NEW(): Tree =  { // careful, map Int to Int

    val hmap = dfa.deltaq( 0 ); // all the initial states

    var i = 0;
    val n = hmap.keySet().size(); // all transitions defined

    val tmapTag = new TreeMap();
    val  tmapBody = new TreeMap();
    var  it = hmap.keySet().iterator();
    while(it.hasNext()) {
      val  targetL = it.next().asInstanceOf[Integer];
      val targetR  = hmap.get( targetL ).asInstanceOf[Integer];

      val I = new Integer( i );
      tmapTag.put( targetL, I );
      tmapBody.put( I, callFun( Predef.Array[Tree] (
        cf.SeqTrace_tail( _iter() ),
        gen.mkIntLit( cf.pos, targetR.intValue() ) )));
      i = i + 1;
    }
    i = 0;
    val tags    = new Array[Int]( n );
    val targets = new Array[Tree]( n );
    var jt = tmapTag.keySet().iterator();
    while(jt.hasNext()) {
      val tagI = jt.next().asInstanceOf[Integer];
      tags( i )    = tagI.intValue();
      val I    = tmapTag.get( tagI ).asInstanceOf[Integer];
      targets( i ) = tmapBody.get( I ).asInstanceOf[Tree];;
      i = i + 1
    }
    gen.Switch( gen.Ident( pos, targetSym ),
               tags,
               targets,
               code_error()/*cannot happen*/ );

  }

  override def currentMatches(label: Label): Tree = {
    label.match {
      case LPair( target, theLab ) =>
        cf.Equals( gen.mkIntLit( cf.pos, target.intValue() ),
                         current() );
      case _ => throw new ApplicationError("expected Pair label");
    }
  }


  override def code_state_NEW(i: Int): Tree = { // precondition i != 0
    var stateBody = code_delta( i, DefaultLabel() );
    if( stateBody == null ) {
      stateBody = code_error();
    }
    val trans = dfa.deltaq( i );
    val tmapTag = new TreeMap();
    val tmapBody = new TreeMap();
    var  j = 0;
    var  labs = dfa.labels().iterator();
    while(labs.hasNext())  {
      val label  = labs.next();
      val next   = trans.get( label ).asInstanceOf[Integer];
      val action = code_delta( i, label.asInstanceOf[Label] );

      if( action != null ) {
        val J = new Integer( j );
        tmapTag.put( label.asInstanceOf[LPair].state, J );
        tmapBody.put( J, action );

        stateBody = gen.If( currentMatches( label.asInstanceOf[Label] ),
                           action,
                           stateBody);
      }
      j = j + 1;
    }
    val n = tmapTag.keySet().size();
    j = 0;
    val tags    = new Array[int]( n );
    val targets = new Array[Tree]( n );
    val it = tmapTag.keySet().iterator();
    while(it.hasNext()) {
      val tagI     = it.next().asInstanceOf[Integer];
      tags( j )    = tagI.intValue();
      val J        = tmapTag.get( tagI ).asInstanceOf[Integer];
      targets( j ) = tmapBody.get( J ).asInstanceOf[Tree];
      j = j + 1;
    }
    if( n > 0 )
      gen.Switch( gen.Ident( pos, targetSym ), tags, targets, code_error() );
    else
      code_error();
  }

    // calling the AlgebraicMatcher here
  override def _cur_match(pat1: Tree): Tree = {
    var pat = pat1;
    //System.out.println("RTiS._cur_match("+pat.toString()+")");
    //System.out.println("calling algebraic matcher on type:"+pat.type);

    //System.err.println( "curT"+currentElem().type().widen() );
    val m = new PartialMatcher( owner, //funSym,//this.funSym,
                               currentElem(),
                               defs.boolean_TYPE() );

    val freshenMap = new HashMap(); // sym2exp -> new sym
    val helpMap3 = new HashMap();         // new sym -> original sym

    // "freshening": never use the same symbol more than once
        // (in later invocations of _cur_match)

    var it = varsToExport.iterator();
    while(it.hasNext() ) {
      val key = it.next().asInstanceOf[Symbol];
      if( key.name.toString().indexOf("$") == -1 ) {
        this.helpMap2.put( key, helpMap.get( key ));
        // "freshening" by appending string ( a bit dangerous )
        val newSym = key.cloneSymbol().setOwner( owner /*funSym*/ );
        newSym.name = Name.fromString( key.name.toString() + "%" );
        freshenMap.put( key, newSym );
        helpMap3.put( newSym, helpMap.get( key ));
        //System.out.println( "key: "+ key + " key.owner:"+key.owner());
        //System.out.println( "newsym owner:"+newSym.owner());
      } else {
        freshenMap.put( key, key );
      }
    }

    //System.out.println("RightTracerInScala:: -pat :"+pat.toString());
    /*
     System.out.println("RightTracerInScala - the seqVars"+seqVars);
     System.out.println("RightTracerInScala - the varsToExport"+varsToExport);
     */
    //System.out.println("RightTracerInScala::freshenMap :"+freshenMap);

        // "freshening"

        /* TEST
         */
    val tc = new TreeCloner( cf.unit.global, freshenMap, Type.IdMap );

        /*
        Transformer tc = new Transformer(cf.unit.global) {
                public Tree transform(Tree tree) {
                    tree = super.transform(tree);
                    if (tree.hasSymbol()) {
                        Object symbol = freshenMap.get(tree.symbol());
                        if (symbol != null) tree.setSymbol((Symbol)symbol);
                    }
                    return tree;
                }
            };
            */
    pat = tc.transform( pat );


        //System.out.println("RightTracerInScala:: -pat( after subst ) :"+pat);


        // val match { case <pat> => { <do binding>; true }
        //             case _     => false


    val ts = new Array[Tree]( helpMap3.keySet().size() );
    var j = 0;
    var jt = helpMap3.keySet().iterator();
    while(jt.hasNext()) {
      val vsym = jt.next().asInstanceOf[Symbol];
      val hv   = helpMap3.get( vsym ).asInstanceOf[Symbol];
      //hv.setType( defs.LIST_TYPE( elementType ) ) ; DEBUG ALARM ?
      val refv   = gen.Ident(cf.pos, vsym);
      val refhv  = gen.Ident(cf.pos, hv);
      ts( j ) = gen.Assign( refhv, refv );
      j = j + 1;
            // System.out.println( "the assign" + res[ j - 1 ] );
    }

    val res = gen.mkBooleanLit( cf.pos, true ); // just `true'
    val theBody =  gen.mkBlock(ts, res);

    am.construct( m, Predef.Array[Tree] (

      cf.gen.CaseDef( pat,           // if tree val matches pat -> update vars, return true
                     theBody/* "freshening */),
      cf.gen.CaseDef( cf.gen.Ident( pat.pos, defs.PATTERN_WILDCARD ),
                     gen.mkBooleanLit( pat.pos, false )) ), // else return false
                 true // do binding please
               );

    am.toTree();
  }

  /** returns translation of transition with label from i.
   *  returns null if there is no such transition(no translation needed)
   */
  override def code_delta(i: Int , label: Label  ): Tree = {
    val hmap    = dfa.deltaq( i );
    val ntarget =  hmap.get( label ).asInstanceOf[Integer];
    var algMatchTree: Tree = null;
    if( ntarget == null )
      return null;

    //System.out.println("delta("+i+","+label+")" );
    var  theLab: Label = null;
    label.match {
      case LPair ( state, lab2 )=>
        //assert ntarget == state;
        theLab = lab2;
      lab2.match {
        case TreeLabel( pat ) =>
          algMatchTree = _cur_match( pat );
        case _ =>
      }
      case DefaultLabel() =>
        throw new ApplicationError(); // should not happen
    }
    //assert dfa.qbinders != null : "qbinders ?";

    var vars = dfa.qbinders( i );

    //System.out.println("dfa.qbinders[ i ]"+vars);

    if( null == vars ) vars = new Vector(); // TODO: make this more consistent
    //assert vars != null;

    val stms = new Array[Tree]( vars.size()
                                + { if (algMatchTree != null ) 1 else 0 });
    var j = 0;
    var it = vars.iterator();
    while(it.hasNext()) {
      val vble = it.next().asInstanceOf[Symbol];
      val rhs = gen.Ident( pos, curSym );
      stms( j ) = prependToHelpVar( vble , rhs);
      j = j + 1;
    }

    if( algMatchTree != null ) {
      stms( j ) = algMatchTree ;
      j = j + 1
    }

    val value = callFun( Predef.Array[Tree] ( cf.SeqTrace_tail( _iter() ),
                                      gen.mkIntLit( cf.pos, ntarget.intValue() ) ) );

    gen.mkBlock( pos, stms, value );
  }

  override def stateWrap(i: Int): Tree = {
    if( i == 0 )
      code_state0_NEW();
    else
      code_state_NEW( i );
  }

  /* returns statements that do the work of the right-transducer
   */
  def getStms(trace: Tree, unit: CompilationUnit, body: Tree): Tree =  {

    val stms = new TreeList();
    val loopbody = code_body_NEW();

    stms.append( gen.ValDef( iterSym, trace ) );
    stms.append( gen.ValDef( stateSym, gen.mkIntLit( cf.pos, 0 ) ) );
    stms.append( helpVarDefs );
    stms.append( gen.LabelDef( this.funSym,
                              Predef.Array[Ident] (
                                gen.Ident( pos, iterSym ),
                                gen.Ident( pos, stateSym )
                              ), loopbody ));

    // bind variables handled by this righttracer
    var it = seqVars.iterator();
    while(it.hasNext()) {
      stms.append( bindVar( it.next().asInstanceOf[Symbol] ) );
    }

    val treeCloner = new Transformer(unit.global) {
      override def transform(tree1: Tree): Tree = {
        val tree = super.transform(tree1);
        if (tree.hasSymbol()) {
          val symbol = helpMap2.get(tree.symbol());
          if (symbol != null) tree.setSymbol(symbol.asInstanceOf[Symbol]);
        }
        tree;
      }
    };

    gen.mkBlock(stms.toArray(), treeCloner.transform( body ));
  }



  /** return the accumulator. (same as in LeftTracerInScala)
   *  todo: move tree generation of Unit somewhere else
   */
   override def run_finished(state: Int):  Tree = {
     gen.mkUnitLit(Position.FIRSTPOS);
    }

  def   current() = {   gen.Ident( pos, targetSym );}

    def refHelpVar(realVar: Symbol) = {
      val hv = helpMap.get( realVar ).asInstanceOf[Symbol];
      //assert hv != null : realVar;
      gen.Ident(Position.FIRSTPOS, hv);
    }

  def assignToHelpVar(realVar: Symbol, rhs: Tree): Tree = {
    val hv = refHelpVar( realVar );
    gen.Assign( hv, rhs );
  }

  def bindVar(realVar: Symbol): Tree = {
    val hv = refHelpVar( realVar );
        /*
          System.out.println("binding realVar.name "+realVar.name+" type:"+realVar.type()+" to hv type:"+hv.type());
          realVar.setOwner( owner );
          System.out.println("is same as realVar"+realVar.type().isSameAs( elementType ));
          System.out.println("is same as hv"+realVar.type().isSameAs( hv.type() ));
          if( realVar.type().isSameAs( elementType ))
          return gen.ValDef( realVar, cf.SeqList_head( hv ));
          else
          return gen.ValDef( realVar, hv );
        */
    if( realVar.getType().isSameAs( hv.getType())) {
      gen.ValDef( realVar, hv ); // e.g. x @ _*
    } else
      gen.ValDef( realVar, cf.SeqList_head( hv ));
  }
}
}

File: sources/scalac/transformer/matching/RightTracerInScala.java
package scalac.transformer.matching ;

import scalac.*;
import scalac.ast.*;
import scalac.symtab.*;
import Tree.*;

import scalac.transformer.TransMatch.Matcher ;

import java.util.* ;
import Scope.SymbolIterator;

import scalac.ast.printer.TextTreePrinter ;

import scalac.util.Name ;
import scalac.util.Names ;

import ch.epfl.lamp.util.Position;

public class RightTracerInScala extends Autom2Scala  {

      //Scope  scp;
      //Symbol vars[];

      Vector seqVars;
      Vector allVars;

      Matcher _m;

      /** translate right tracer to code
       * @param dfa determinized left tracer
       * @param left nondeterm. left tracer
       * @param cf   ...
       * @param pat  ?
       * @param elementType ...
       */
      public RightTracerInScala( DetWordAutom dfa,
                                 NondetWordAutom left,
                                 Matcher m,
                                 CodeFactory cf,
                                 Tree pat,
                                 Type elementType ) {
            super( dfa, elementType, m.owner, cf );
            this._m = m;


            Vector seqVars = new Vector();

            for( int j = 0; j < left.nstates; j++ ) {
                  if( left.qbinders[ j ] != null )
                        for( Iterator it = left.qbinders[ j ].iterator();
                             it.hasNext() ; ) {
                              Symbol varSym = (Symbol) it.next();

                              if( !seqVars.contains( varSym ) )
                                    seqVars.add( varSym );
                        }
            }

            this.seqVars = seqVars;
            this.allVars = CollectVariableTraverser.collectVars( pat );

            helpMap = new HashMap();
            helpMap2 = new HashMap();
            helpVarDefs = new Vector();

            for( Iterator it = seqVars.iterator(); it.hasNext(); ) {
                  makeHelpVar( (Symbol) it.next() );
            }

            for( Iterator it = allVars.iterator(); it.hasNext(); ) {
                  Symbol varSym = (Symbol) it.next();
                  if( !seqVars.contains( varSym )) {
                        makeHelpVar( varSym, true );
                  }
            }

            initializeSyms();
      }

      //      Symbol funSym;

      Symbol elemSym;
      Symbol targetSym;

      HashMap helpMap ;
      HashMap helpMap2 ;
      Vector  helpVarDefs;

      void makeHelpVar( Symbol realVar ) {
            makeHelpVar( realVar, false );
      }

      void makeHelpVar( Symbol realVar, boolean keepType ) {
	  Symbol helpVar = new TermSymbol( //Kinds.VAR,
                                             pos,
                                             cf.fresh.newName( realVar.name
                                                               .toString() ),
                                             owner,
                                             0);

            //System.out.println("making helpvar : "+realVar+" -> "+helpVar);

            if( keepType )
                  helpVar.setType( realVar.type() );
            else
                  helpVar.setType( cf.SeqListType( elementType ) );


            helpMap.put( realVar, helpVar );

            Tree rhs;
            if( keepType )
                  rhs = cf.ignoreValue( realVar.type() );
            else
		rhs = /* cf.newRef( */ cf.newSeqNil( elementType ) /* ) */;
	    helpVar.flags |= Modifiers.MUTABLE;
            Tree varDef = gen.ValDef(helpVar, rhs);
            //((ValDef) varDef).kind = Kinds.VAR;
            helpVarDefs.add( varDef );

      }

      Tree refHelpVar( Symbol realVar ) {
            Symbol hv = (Symbol)helpMap.get( realVar );
            assert hv != null : realVar;
            return gen.Ident(Position.NOPOS, hv);
      }

      Tree prependToHelpVar( Symbol realVar, Tree elem ) {
            Tree hv = refHelpVar( realVar );
            return gen.Assign( hv, cf.newSeqCons( elem, hv ));
            /*
            return cf.Block(pos,
                            new Tree [] {
                  cf.debugPrintRuntime( "ASSIGN" ),
                  gen.Assign( hv, cf.newSeqCons( elem, hv ))
                        }, defs.UNIT_TYPE);
            */
      }

      Tree bindVar(Symbol realVar) {
            Tree hv = refHelpVar( realVar );
            //System.out.println("binding realVar.name "+realVar.name+" type:"+realVar.type()+" to smth");
            realVar.setOwner( owner );
            if( realVar.type().isSameAs( elementType ))
                  return gen.ValDef( realVar, cf.SeqList_head( hv ));
            else
                  return gen.ValDef( realVar, hv);

      }

      protected void initializeSyms() {

            this.funSym = newFunSym( "binder" );

            this.iterSym = new TermSymbol( //Kinds.VAL,
                                           pos,
                                           cf.fresh.newName("iter"),
                                           funSym,
                                           0)
                  .setType( cf.SeqTraceType( elementType ));

            this.stateSym = new TermSymbol( //Kinds.VAL,
                                            pos,
                                            cf.fresh.newName("q"),
                                            funSym,
                                            0 )
                  .setType( defs.INT_TYPE ) ;

            this.elemSym = new TermSymbol( //Kinds.VAL,
                                           pos,
                                           cf.fresh.newName("elem"),
                                           funSym,
                                           0)
                  .setType( elementType ) ;

            this.targetSym = new TermSymbol( //Kinds.VAL,
                                             pos,
                                             cf.fresh.newName("trgt"),
                                             funSym,
                                             0)
                  .setType( defs.INT_TYPE ) ;


            funSym.setType( new Type.MethodType( new Symbol[] {
                  iterSym, stateSym },  defs.UNIT_TYPE ));

      }

      // same as in LeftTracer
      Tree code_fail() {

            return cf.ThrowMatchError( _m.pos, defs.UNIT_TYPE );

      }

      public Tree code_body() {

            Tree body = code_fail(); // never reached at runtime.

            // state [ nstates-1 ] is the dead state, so we skip it

            //`if( state == q ) <code_state> else {...}'
            for( int i = dfa.nstates-1; i >= 0; i-- ) {
                  body = code_state( i, body );
            }


            Tree t3 = cf.If( cf.isEmpty( _iter() ),
                               run_finished( 0 ),
                               gen.Block( new Tree[] {
                                     gen.ValDef( targetSym,
                                                 cf.SeqTrace_headState( gen.Ident( pos, iterSym))),
                                     gen.ValDef( elemSym,
                                                 cf.SeqTrace_headElem( gen.Ident( pos, iterSym))),

                                     body }));

            /*
            t3 = gen.Block( new Tree[] {
                  cf.debugPrintRuntime("enter binderFun"),
                  cf.debugPrintRuntime(" state:"),
                  cf.debugPrintRuntime( gen.Ident( pos, stateSym )),
                  cf.debugPrintRuntime(" iter:"),
                  cf.debugPrintRuntime(_iter()),
                  cf.debugPrintNewlineRuntime(""),
                  t3 });
            */

            //System.out.println("enter RightTracerInScala:code_body()");// DEBUG
            //System.out.println("dfa.nstates"+dfa.nstates);// DEBUG
            return t3;
      }

      /** this one is special, we check the first element of the trace
       *  and choose the next state depending only on the state part
       */
      Tree code_state0( Tree elseBody ) { // careful, map Int to Int

            HashMap hmap    = (HashMap) dfa.deltaq( 0 ); // all the initial states

            Tree stateBody = code_fail(); // never happens

            for( Iterator it = hmap.keySet().iterator(); it.hasNext(); ) {
                  Integer targetL  = (Integer) it.next();
                  Integer targetR  = (Integer) hmap.get( targetL );

                  stateBody = cf.If( cf.Equals( gen.Ident( pos, targetSym ),
                                                cf.Int( targetL )),
                                     callFun( new Tree[] {
                                           cf.SeqTrace_tail( _iter() ),
                                           cf.Int( targetR ) }),
                                     stateBody );
            }

            return cf.If( cf.Equals( _state(), cf.Int( 0 )),
                          stateBody ,
                          elseBody );

      }

      Tree code_state( int i, Tree elseBody ) {

            if( i == 0 )
                  return code_state0( elseBody );

            int  finalSwRes;
            Tree stateBody ; // action(delta) for one particular label/test

            // default action (fail if there is none)

            stateBody = code_delta( i, Label.DefaultLabel);

            if( stateBody == null )
                  stateBody = code_fail();

            // transitions of state i

            HashMap trans = ((HashMap[])dfa.deltaq)[ i ];

            for( Iterator labs = dfa.labels.iterator(); labs.hasNext() ; ) {
                  Object label = labs.next();
                  Integer next = (Integer) trans.get( label );

                  Tree action = code_delta( i, (Label) label );

                  if( action != null ) {

                        stateBody = cf.If( _cur_eq( _iter(), (Label) label ),
                                           action,
                                           stateBody);
                  }
            }

            return wrapStateBody0( stateBody,
                                   elseBody,
                                   i );
      }

      /** returns a Tree whose type is boolean.
       *  now we care about free vars
       */

      Tree handleBody( Object o ) {
            HashMap helpMap = (HashMap) o;
            Tree res[] = new Tree[ helpMap.keySet().size() + 1 ];
            int j = 0;
            for( Iterator it = helpMap.keySet().iterator(); it.hasNext(); ) {
                  Symbol vsym = (Symbol) it.next();
                  Symbol hv   = (Symbol) helpMap.get( vsym );
                  hv.setType( cf.SeqListType( elementType ) ) ;
                  Tree refv   = gen.Ident(Position.NOPOS, vsym);
                  Tree refhv  = gen.Ident(Position.NOPOS, hv);
                  res[ j++ ] = gen.Assign( refhv, refv );
            }

            res[ j ] = super.handleBody( freeVars ); // just `true'

            return cf.Block(Position.NOPOS, res, res[j].type() );
      }

      // calling the AlgebraicMatcher here
      Tree _cur_match( Tree pat ) {

            //System.out.println("RTiS._cur_match("+TextTreePrinter.toString(pat)+")");

            //System.out.println("calling algebraic matcher on type:"+pat.type);

            Matcher m = new Matcher( funSym,//this.funSym,
                                     currentElem(),
                                     defs.BOOLEAN_TYPE );

            Vector varsToExport = NoSeqVariableTraverser.varsNoSeq( pat );

            //System.out.println("RightTracerInScala::varsToExport :"+varsToExport);

            for( Iterator it = varsToExport.iterator(); it.hasNext(); ) {
                  Symbol key = (Symbol) it.next();
                  this.helpMap2.put( key, helpMap.get( key ));
            }

	    // find this weird ? pattern matcher expects var. symbol for _ pattern FIXME ?!! CANNOT BE TRUE
	    //Symbol obfuscvble = new TermSymbol(0, Name.fromString("ga$ga$ga$"), _m.owner, 0).setType( pat.type() );

            am.construct( m, new CaseDef[] {
                  (CaseDef) cf.make.CaseDef( pat.pos,
                                             pat,
                                             Tree.Empty,
                                             handleBody( helpMap2 )),
                        (CaseDef) cf.make.CaseDef( pat.pos,
                                                   cf.make.Ident(pat.pos, Names.WILDCARD)
                                                   //.setSymbol( Symbol.NONE ) FIXED
						   .setType( pat.type() ),
                                                   Tree.Empty,
                                                   gen.mkBooleanLit( pat.pos, false )) },
                          true // do binding please
                          );
            Tree res = am.toTree().setType( defs.BOOLEAN_TYPE );
            //System.out.println("freeVars: "+freeVars);
            return res;
      }


      /** returns translation of transition with label from i.
       *  returns null if there is no such transition(no translation needed)
       */
      Tree code_delta( int i, Label label ) {
            HashMap hmap    = (HashMap) dfa.deltaq( i );
            Integer ntarget = (Integer) hmap.get( label );
            Tree    algMatchTree = null;
            if( ntarget == null )
                  return null;
            //System.out.println("delta("+i+","+label+")" );
            Label theLab = null;
            switch(label) {
            case Label.Pair( Integer state, Label lab2 ):
                  //assert ntarget == state;
                  theLab = lab2;
                  switch( lab2 ) {
                  case TreeLabel( Tree pat ):
                        algMatchTree = _cur_match( pat );
                        break;
                  }
                  break;
            case DefaultLabel:
                  throw new ApplicationError(); // should not happen
            }
            assert dfa.qbinders != null : "qbinders ?";

            Vector vars = dfa.qbinders[ i ];

            if( vars == null ) vars = new Vector(); // TODO: make this more consistent
            assert vars != null;

            //System.out.println("delta: theLab: " + theLab + " vars in current ="+ vars );

            if( ntarget == null )
                  return code_fail();

            Tree stms[] = new Tree[ vars.size()
                                    + ((algMatchTree != null )? 1 : 0 )
                                    + 1 ];
            int j = 0;
            for( Iterator it = vars.iterator(); it.hasNext(); ) {
                  Symbol var = (Symbol) it.next();

                  /*Tree field = gen.Select( gen.Ident( pos, accumSym ),
                                           var.name ).setSymbol( var );
                  */
                  //Tree field = gen.Ident( pos, var );
                  Tree rhs = gen.Ident( pos, elemSym ); //cf._cur( _iter() );

                  stms[ j++ ] = prependToHelpVar( var , rhs);
            }

            if( algMatchTree != null )
                  stms[ j++ ] = algMatchTree ;

            stms[ j ] = callFun( new Tree[] { cf.SeqTrace_tail( _iter() ),
                                              cf.Int( ntarget ) } );

            return cf.Block( pos, stms, funRetType() );
      }

      /* returns statements that do the work of the right-transducer
       */
      Tree[] getStms( Tree trace ) {

            //System.out.println( "!!getStms.helpVarDefs: "+helpVarDefs);

            Vector v = new Vector();

            for( Iterator it = helpVarDefs.iterator(); it.hasNext(); )
                  v.add( (Tree) it.next() );

            v.add( gen.DefDef( this.funSym, code_body() )  );
            v.add( callFun( new Tree[] {  trace, cf.Int( 0 )  }  )  );


            /*
            for(Iterator it = helpMap.keySet().iterator(); it.hasNext(); ) {
                  // DEBUG
                  Symbol var = (Symbol)it.next();
                  v.add( cf.debugPrintRuntime( var.name.toString() ));
                  v.add( cf.debugPrintRuntime( refHelpVar( var )) );
            }
            */

            for( Iterator it = seqVars.iterator(); it.hasNext(); ) {
                  v.add( bindVar( (Symbol) it.next() ) );
            }

            Tree result[] = new Tree[ v.size() ];
            int j = 0;
            for( Iterator it = v.iterator(); it.hasNext(); ) {
                  result[ j++ ] = (Tree) it.next();
            }

            // helpvarSEQ via _m.varMap

            return result;

      }

      /** return the accumulator. (same as in LeftTracerInScala)
       */
      Tree run_finished( int state ) {
            //return gen.Ident( accumSym.pos, accumSym );
	  return gen.Block(0, Tree.EMPTY_ARRAY).setType( defs.UNIT_TYPE );
	      /* gen.Ident(0, defs.NULL); */
	  /*gen.New( pos, defs.SCALA_TYPE, defs.NULL, //UNIT_CLASS,
                            Type.EMPTY_ARRAY,
                            Tree.EMPTY_ARRAY);                             */


      }

      Tree current() {
            return gen.Ident( pos, targetSym );
      }

      Tree currentElem() {
            return gen.Ident( pos, elemSym );
      }

      Tree _cur_eq( Tree iter, Label label ) {
            //System.out.println("_cur_eq, thelab: "+label);
            switch( label ) {
            case Pair( Integer target, Label theLab ):
                  return cf.Equals( cf.Int( target ),
                                    current() );
            }
            throw new ApplicationError("expected Pair label");
      }


}
